circuit/src/subquery/caller.rs

```diff
@@ -154,4 +173,68 @@ impl<P: JsonRpcClient, F: Field> SubqueryCaller<P, F> {
         };
         hilo
     }
+
+    pub fn groth16_verify(
+        &mut self,
+        ctx: &mut Context<F>,
+        range: &RangeChip<F>,
+        input: Groth16AssignedInput<F>,
+    ) -> HiLo<AssignedValue<F>> {
+        let constants = get_groth16_consts_from_max_pi(self.groth16_max_pi);
+        let mut fe = input.vkey_bytes;
+        fe.extend(input.proof_bytes);
+        fe.extend(input.public_inputs);
+        assert_eq!(fe.len(), constants.num_fe_per_input);
+        let zero = ctx.load_witness(F::ZERO);
+        fe.resize_with(constants.max_num_fe_per_input_no_hash, || zero);
+        let to_hash = fe
+            .iter()
+            .map(|v| *v.value())
+            .collect::<Vec<_>>()
+            .chunks(NUM_FE_PER_CHUNK)
```

Comment:
this is still hardcoded for max_pi=4?

Reply:
or it's just set by the max subquery input length allowed?

Reply:
NUM_FE_PER_CHUNK is just the max subquery input length allowed (13)